File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/ui/calendar/dialogs/MonthOverviewDialog.java
package com.byagowi.persiancalendar.ui.calendar.dialogs;

import android.app.Dialog;
import android.content.Context;
import android.os.Bundle;
import android.text.TextUtils;
import android.util.SparseArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import com.byagowi.persiancalendar.R;
import com.byagowi.persiancalendar.calendar.AbstractDate;
import com.byagowi.persiancalendar.databinding.MonthOverviewDialogBinding;
import com.byagowi.persiancalendar.databinding.MonthOverviewItemBinding;
import com.byagowi.persiancalendar.entities.AbstractEvent;
import com.byagowi.persiancalendar.entities.DeviceCalendarEvent;
import com.byagowi.persiancalendar.utils.CalendarType;
import com.byagowi.persiancalendar.utils.Utils;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.google.android.material.bottomsheet.BottomSheetDialogFragment;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class MonthOverviewDialog extends BottomSheetDialogFragment {
    private static String BUNDLE_KEY = "jdn";

    public static MonthOverviewDialog newInstance(long jdn) {
        Bundle args = new Bundle();
        args.putLong(BUNDLE_KEY, jdn);

        MonthOverviewDialog fragment = new MonthOverviewDialog();
        fragment.setArguments(args);
        return fragment;
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Bundle args = getArguments();
        Context context = getContext();

        long baseJdn = args == null ? -1 : args.getLong(BUNDLE_KEY, -1);
        if (baseJdn == -1) baseJdn = Utils.getTodayJdn();

        List<MonthOverviewRecord> records = new ArrayList<>();

        CalendarType mainCalendar = Utils.getMainCalendar();
        AbstractDate date = Utils.getDateFromJdnOfCalendar(mainCalendar, baseJdn);
        long monthLength = Utils.getMonthLength(mainCalendar, date.getYear(), date.getMonth());
        SparseArray<List<DeviceCalendarEvent>> deviceEvents = Utils.readMonthDeviceEvents(context, baseJdn);
        for (long i = 0; i < monthLength; ++i) {
            long jdn = baseJdn + i;
            List<AbstractEvent> events = Utils.getEvents(jdn, deviceEvents);
            String holidays = Utils.getEventsTitle(events, true, false, false, false);
            String nonHolidays = Utils.getEventsTitle(events, false, false, true, false);
            if (!(TextUtils.isEmpty(holidays) && TextUtils.isEmpty(nonHolidays)))
                records.add(new MonthOverviewRecord(Utils.dayTitleSummary(
                        Utils.getDateFromJdnOfCalendar(mainCalendar, jdn)), holidays, nonHolidays));
        }
        if (records.size() == 0)
            records.add(new MonthOverviewRecord(getString(R.string.warn_if_events_not_set), "", ""));

        MonthOverviewDialogBinding binding = MonthOverviewDialogBinding.inflate(
                LayoutInflater.from(context), null, false);
        binding.recyclerView.setLayoutManager(new LinearLayoutManager(context));
        binding.recyclerView.setAdapter(new ItemAdapter(records));

        BottomSheetDialog bottomSheetDialog = new BottomSheetDialog(context);
        bottomSheetDialog.setContentView(binding.getRoot());
        bottomSheetDialog.setCancelable(true);
        bottomSheetDialog.setCanceledOnTouchOutside(true);
        return bottomSheetDialog;
    }

    static class MonthOverviewRecord {
        final String title;
        final String holidays;
        final String nonHolidays;

        MonthOverviewRecord(String title, String holidays, String nonHolidays) {
            this.title = title;
            this.holidays = holidays;
            this.nonHolidays = nonHolidays;
        }
    }

    private class ItemAdapter extends RecyclerView.Adapter<ItemAdapter.ViewHolder> {
        private List<MonthOverviewRecord> mRows;

        ItemAdapter(List<MonthOverviewRecord> rows) {
            mRows = rows;
        }

        @NonNull
        @Override
        public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            MonthOverviewItemBinding binding = MonthOverviewItemBinding.inflate(
                    LayoutInflater.from(parent.getContext()), parent, false);

            return new ViewHolder(binding);
        }

        @Override
        public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
            holder.bind(position);
        }

        @Override
        public int getItemCount() {
            return mRows.size();
        }

        class ViewHolder extends RecyclerView.ViewHolder {
            MonthOverviewItemBinding mBinding;

            ViewHolder(@NonNull MonthOverviewItemBinding binding) {
                super(binding.getRoot());
                mBinding = binding;
            }

            public void bind(int position) {
                MonthOverviewRecord record = mRows.get(position);
                mBinding.title.setText(record.title);
                mBinding.holidays.setText(record.holidays);
                mBinding.holidays.setVisibility(TextUtils.isEmpty(record.holidays) ? View.GONE : View.VISIBLE);
                mBinding.nonHolidays.setText(record.nonHolidays);
                mBinding.nonHolidays.setVisibility(TextUtils.isEmpty(record.nonHolidays) ? View.GONE : View.VISIBLE);
            }
        }
    }
}

File: PersianCalendar/src/main/java/com/byagowi/persiancalendar/ui/calendar/dialogs/MonthOverviewDialog.kt
package com.byagowi.persiancalendar.ui.calendar.dialogs

import android.app.Dialog
import android.os.Bundle
import android.view.View
import android.view.ViewGroup
import androidx.core.os.bundleOf
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.byagowi.persiancalendar.R
import com.byagowi.persiancalendar.databinding.MonthOverviewDialogBinding
import com.byagowi.persiancalendar.databinding.MonthOverviewItemBinding
import com.byagowi.persiancalendar.utils.Jdn
import com.byagowi.persiancalendar.utils.copyToClipboard
import com.byagowi.persiancalendar.utils.dayTitleSummary
import com.byagowi.persiancalendar.utils.dp
import com.byagowi.persiancalendar.utils.getEvents
import com.byagowi.persiancalendar.utils.getEventsTitle
import com.byagowi.persiancalendar.utils.getMonthLength
import com.byagowi.persiancalendar.utils.isHighTextContrastEnabled
import com.byagowi.persiancalendar.utils.layoutInflater
import com.byagowi.persiancalendar.utils.mainCalendar
import com.byagowi.persiancalendar.utils.readMonthDeviceEvents
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.bottomsheet.BottomSheetDialogFragment

class MonthOverviewDialog : BottomSheetDialogFragment() {

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val activity = requireActivity()

        val baseJdn = arguments?.getLong(BUNDLE_KEY, -1L)?.takeIf { it != -1L }?.let(::Jdn) ?: Jdn.today
        val date = baseJdn.toCalendar(mainCalendar)
        val deviceEvents = readMonthDeviceEvents(activity, baseJdn)
        val events = (0 until getMonthLength(mainCalendar, date.year, date.month)).mapNotNull {
            val jdn = baseJdn + it
            val events = getEvents(jdn, deviceEvents)
            val holidays = getEventsTitle(
                events, holiday = true, compact = false, showDeviceCalendarEvents = false,
                insertRLM = false, addIsHoliday = isHighTextContrastEnabled
            )
            val nonHolidays = getEventsTitle(
                events, holiday = false, compact = false, showDeviceCalendarEvents = true,
                insertRLM = false, addIsHoliday = false
            )
            if (holidays.isEmpty() && nonHolidays.isEmpty()) null
            else MonthOverviewRecord(
                dayTitleSummary(jdn.toCalendar(mainCalendar)), holidays, nonHolidays
            )
        }.takeIf { it.isNotEmpty() } ?: listOf(
            MonthOverviewRecord(getString(R.string.warn_if_events_not_set), "", "")
        )

        return BottomSheetDialog(activity).apply {
            setContentView(
                MonthOverviewDialogBinding.inflate(
                    activity.layoutInflater, null, false
                ).apply {
                    recyclerView.apply {
                        layoutManager = LinearLayoutManager(context)
                        adapter = ItemAdapter(events)
                        setPadding(0, 4.dp, 0, 0)
                    }
                }.root
            )
            setCancelable(true)
            setCanceledOnTouchOutside(true)
        }
    }

    internal class MonthOverviewRecord(
        val title: String, val holidays: String, val nonHolidays: String
    ) {
        override fun toString() = listOf(title, holidays, nonHolidays)
            .filter { it.isNotEmpty() }.joinToString("\n")
    }

    private inner class ItemAdapter(private val rows: List<MonthOverviewRecord>) :
        RecyclerView.Adapter<ItemAdapter.ViewHolder>() {

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int) = ViewHolder(
            MonthOverviewItemBinding.inflate(parent.context.layoutInflater, parent, false)
        )

        override fun onBindViewHolder(holder: ViewHolder, position: Int) = holder.bind(position)

        override fun getItemCount(): Int = rows.size

        inner class ViewHolder(private val binding: MonthOverviewItemBinding) :
            RecyclerView.ViewHolder(binding.root), View.OnClickListener {

            init {
                binding.root.setOnClickListener(this)
            }

            fun bind(position: Int) = binding.run {
                val record = rows[position]
                title.text = record.title
                holidays.text = record.holidays
                holidays.visibility = if (record.holidays.isEmpty()) View.GONE else View.VISIBLE
                nonHolidays.text = record.nonHolidays
                nonHolidays.visibility =
                    if (record.nonHolidays.isEmpty()) View.GONE else View.VISIBLE
            }

            override fun onClick(v: View?) = copyToClipboard(
                binding.root, "Events", rows[bindingAdapterPosition].toString(),
                showToastInstead = true
            )
        }
    }

    companion object {
        private const val BUNDLE_KEY = "jdn"

        fun newInstance(jdn: Long) = MonthOverviewDialog().apply {
            arguments = bundleOf(BUNDLE_KEY to jdn)
        }
    }
}
